### PR TITLE
doc: 3025 http keyword updates v3

### PR DESCRIPTION
**Pull request by jmtaylor90**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/3025

Describe changes:
- applied feedback from #10101 
- updated all http keywords documentation
- RTD build of branch - https://rtd-test-suricata.readthedocs.io/en/latest/rules/http-keywords.html

### Provide values to any of the below to override the defaults.

To use a pull request use a branch name like `pr/N` where `N` is the
pull request number.

Alternatively, `SV_BRANCH` may also be a link to an
OISF/suricata-verify pull-request.

```
SV_REPO=
SV_BRANCH=
SU_REPO=
SU_BRANCH=
LIBHTP_REPO=
LIBHTP_BRANCH=
```

### CODE REVIEW

**Review by ozuriexv — COMMENTED** [review]
I think it'd be beneficial to add a reference to 'HTTP Primer' explaining what 'sticky buffers' are.  It seems to be explained briefly in '[8.1.6.1 - Modifier Keywords](https://rtd-test-suricata.readthedocs.io/en/latest/rules/intro.html#modifier-keywords)'.

**Comment by jmtaylor90** [comment]
> I think it'd be beneficial to add a reference to 'HTTP Primer' explaining what 'sticky buffers' are. It seems to be explained briefly in '[8.1.6.1 - Modifier Keywords](https://rtd-test-suricata.readthedocs.io/en/latest/rules/intro.html#modifier-keywords)'.

will add in next PR

File: doc/userguide/rules/http-keywords.rst [hunk 1]
```diff
@@ -1,813 +1,1200 @@
...
+  Example"; flow:established,to_client; :example-rule-options:`http.stat_code; \
+  content:"200";` classtype:bad-unknown; sid:117; rev:1;)
 
-Example of the purpose of ``http.stat_code``:
+.. note:: ``http.stat_code`` does not include the leading or trailing space
 
-.. image:: http-keywords/stat-code1.png
+.. _http.stat_msg:
 
 http.stat_msg
```
Comment (zoomequipd) [inline doc/userguide/rules/http-keywords.rst:854]: Is there any chance we can get a note that reflects the behavior of http.stat_msg on HTTP/2 traffic. 

https://redmine.openinfosecfoundation.org/issues/4067#note-37
```
Philippe Antoine wrote in [#note-37](https://redmine.openinfosecfoundation.org/issues/4067#note-37):
>
>The empty buffer is http_stat_msg for HTTP2
```

Commit creating the constant empty buffer: 
https://github.com/OISF/suricata/commit/22fb385e23991f0a7e3b83aeb2693d8ab21bdf41

File: doc/userguide/rules/http-keywords.rst [hunk 2]
```diff
@@ -1,813 +1,1200 @@
...
--  ``http.response_body`` will match on gzip decoded data just like
-   ``file.data`` does.
+  Server response body
 
--  Since ``http.response_body`` matches on a server response, it
-   can't be used with the ``to_server`` or ``from_client`` flow
-   directives.
+.. container:: example-rule
 
--  Corresponding PCRE modifier: ``Q``
```
Comment (jufajardini) [inline doc/userguide/rules/http-keywords.rst:646]: Is it ok to remove this note?

File: doc/userguide/rules/http-keywords.rst [hunk 3]
```diff
@@ -1,813 +1,1200 @@
...
+  content:"suricata.io, oisf.net";` classtype:bad-unknown; sid:125; rev:1;)
+
+.. _http.host.raw:
+
+http.host.raw
+-------------
+
+The ``http.host.raw`` buffer matches on HTTP host name content but does not
+have more than the standard normalization performed on the buffer contents.
+(see :ref:`http.host`)
```
Comment (jufajardini) [inline doc/userguide/rules/http-keywords.rst:1056]: The way this is worded for `http.header.raw` is somewhat easier to follow and understand. Do they mean the same?

File: doc/userguide/rules/http-keywords.rst [hunk 4]
```diff
@@ -1,813 +1,1200 @@
...
+It is possible to use any of the :doc:`payload-keywords` with the
+``http.request_header`` keyword.
 
-``http.response_header`` is a 'sticky buffer'.
+For HTTP2, the header name and value get concatenated by ": " (colon and space).
+The colon and space are commonly noted with the hexadecimal format `|3a 20|`
+within signatures.
 
-``http.response_header`` can be used as ``fast_pattern``.
+To detect if a http2 header name contains a ":" (colon), the keyword
```
Comment (jufajardini) [inline doc/userguide/rules/http-keywords.rst:1111]: nit:
```suggestion
To detect if an HTTP2 header name contains a ":" (colon), the keyword
```

File: doc/userguide/rules/http-keywords.rst [hunk 5]
```diff
@@ -1,813 +1,1200 @@
...
 
--  The content match that ``http.host`` applies to must be all lower
-   case or have the ``nocase`` flag set.
+http.response_header
+--------------------
 
--  ``http.host.raw`` matches the unnormalized buffer so matching
-   will be case-sensitive (unless ``nocase`` is set).
+The ``http.response_header`` keyword is used to match on the name and value
+of a HTTP1 or HTTP2 request.
```
Comment (jufajardini) [inline doc/userguide/rules/http-keywords.rst:1134]: nit:
```suggestion
of an HTTP1 or HTTP2 request.
```

File: doc/userguide/rules/http-keywords.rst [hunk 6]
```diff
@@ -1,813 +1,1200 @@
...
-   Example request::
+For HTTP2, the header name and value get concatenated by ": " (colon and space).
+The colon and space are commonly noted with the hexadecimal format `|3a 20|`
+within signatures.
 
-          GET /test.html HTTP/1.1
-          Host: ABC.com
-          Accept: */*
-          Host: efg.net
+To detect if a http2 header name contains a ":" (colon), the keyword
```
Comment (jufajardini) [inline doc/userguide/rules/http-keywords.rst:1143]: nit:
```suggestion
To detect if an HTTP2 header name contains a ":" (colon), the keyword
```

File: doc/userguide/rules/http-keywords.rst [hunk 7]
```diff
@@ -1,813 +1,1200 @@
...
-rule. This makes it a useful shortcut for applying many content
-matches to the HTTP response body, eliminating the need to modify each
-content match individually.
+  alert http $EXTERNAL_NET any -> $HOME_NET any (msg:"HTTP file.data Example"; \
+  flow:established,to_client; :example-rule-options:`file.data; \
+  content:"Server response body";` classtype:bad-unknown; sid:128; rev:1;)
 
-As the body of a HTTP response can be very large, it is inspected in
-smaller chunks.
+The body of a HTTP responses can be very large, therefore the response body is
```
Comment (jufajardini) [inline doc/userguide/rules/http-keywords.rst:1185]: nit:
```suggestion
The body of HTTP responses can be very large, therefore the response body is
```

File: doc/userguide/rules/http-keywords.rst [hunk 8]
```diff
@@ -1,813 +1,1200 @@
...
-in your :ref:`libhtp configuration section
+in the :ref:`libhtp configuration section
 <suricata-yaml-configure-libhtp>` via the ``response-body-limit``
 setting.
 
-If the HTTP body is a flash file compressed with 'deflate' or 'lzma',
-it can be decompressed and ``file.data`` can match on the decompress data.
-Flash decompression must be enabled under ``libhtp`` configuration:
+.. note:: If the HTTP body is a flash file compressed with 'deflate' or 'lzma',
+  it can be decompressed and ``file.data`` can match on the decompress data.
```
Comment (jufajardini) [inline doc/userguide/rules/http-keywords.rst:1194]: nit: decompressed?

File: doc/userguide/rules/http-keywords.rst [hunk 9]
```diff
@@ -823,38 +1210,46 @@ Flash decompression must be enabled under ``libhtp`` configuration:
...
-
--  file.data is the preferred notation, however, file_data is still
-   recognized by the engine and works as well.
+.. note:: ``file.data`` replaces the previous keyword name, ``file_data``.
+  ``file_data`` can still be used but it is recommended that rules be converted
+  to use ``file.data``.
 
--  If a HTTP body is using gzip or deflate, ``file.data`` will match
-   on the decompressed data.
+.. note:: If a HTTP body is using gzip or deflate, ``file.data`` will match on
```
Comment (jufajardini) [inline doc/userguide/rules/http-keywords.rst:1217]: nit:
```suggestion
.. note:: If an HTTP body is using gzip or deflate, ``file.data`` will match on
```

**Review by jufajardini — CHANGES_REQUESTED** [review]
This was massive, thanks for this undertaking!

Left mostly nit comments for typos and the likes., not very confident to review HTTP workings or rule keyword details...

I wonder if something can be done about the sections that have several notes, as I worry that the sheer amount may render some less visible.

Haven't reviewed changes per commit, so no comments on commit separation logic from me.

Thanks once again for this!

File: doc/userguide/rules/http-keywords.rst [hunk 10]
```diff
@@ -1,813 +1,1200 @@
...
+  content:"200"; bsize:8; http.content_type; content:"text/html"; bsize:9;` \
+  classtype:bad-unknown; sid:30; rev:1;)
+
+Request Keywords:
+ * :ref:`http.uri`
+ * :ref:`http.uri.raw`
+ * :ref:`http.method`
+ * :ref:`http.request_line`
+ * :ref:`http.request_body`
+ * :ref:`http.cookie`
```
[the remark this answers is not in the document]
Comment (jmtaylor90) [reply doc/userguide/rules/http-keywords.rst:62]: Indeed, will update in next PR

File: doc/userguide/rules/http-keywords.rst [hunk 11]
```diff
@@ -1,813 +1,1200 @@
...
-the extracted headers in a single buffer, except for those indicated
-in the documentation that are not able to match by this buffer and
-have their own sticky buffer (e.g. ``http.cookie``). The sticky buffer
-can be used in combination with all previously mentioned content
-modifiers, like ``depth``, ``distance``, ``offset``, ``nocase`` and
-``within``.
+.. note:: If there are multiple values for the same header name, they are
+  concatenated with a comma and space (", ") between each value.
+  More information can be found in RFC 2616
+  `<https://www.rfc-editor.org/rfc/rfc2616.html#section-4.2>`_
```
[the remark this answers is not in the document]
Comment (jmtaylor90) [reply doc/userguide/rules/http-keywords.rst:344]: Yes, I didn't want to hide the external link in this case

File: doc/userguide/rules/http-keywords.rst [hunk 12]
```diff
@@ -1,813 +1,1200 @@
...
+
+To match on non-normalized header data, use the :ref:`http.header.raw` keyword.
+
+Normalization Example::
+
+  GET /index.html HTTP/1.1
+  User-Agent: Mozilla/5.0     \r\n
+  Host: suricata.io
+
+Would be normalized to :example-rule-emphasis:`Mozilla/5.0\r\n`
```
[the remark this answers is not in the document]
Comment (jmtaylor90) [reply doc/userguide/rules/http-keywords.rst:312]: No, will correct that in the next PR

**Comment by jmtaylor90** [comment]
> This was massive, thanks for this undertaking!
> 
> Left mostly nit comments for typos and the likes., not very confident to review HTTP workings or rule keyword details...
> 
> I wonder if something can be done about the sections that have several notes, as I worry that the sheer amount may render some less visible.

Indeed, some of the keywords just have a number of details that are important. Without making the document a wall of text, I figured using the note:: option at least broke up the content a little  but am open to suggestions on that

> Once wondering if we could reorganize the keyword order alphabetically, or semantically, somehow (might be a separate task, if agreed upon).

I was wondering about ordering as well, I would be open to adding to this PR once decided on the ordering method.

> Haven't reviewed changes per commit, so no comments on commit separation logic from me.
> 
> Thanks once again for this!

**Comment by jufajardini** [comment]
> > This was massive, thanks for this undertaking!
> > Left mostly nit comments for typos and the likes., not very confident to review HTTP workings or rule keyword details...
> > I wonder if something can be done about the sections that have several notes, as I worry that the sheer amount may render some less visible.
> 
> Indeed, some of the keywords just have a number of details that are important. Without making the document a wall of text, I figured using the note:: option at least broke up the content a little but am open to suggestions on that
> 

Unfortunately, no suggestions, for now, might be the easiest solution, indeed.

> > Once wondering if we could reorganize the keyword order alphabetically, or semantically, somehow (might be a separate task, if agreed upon).
> 
> I was wondering about ordering as well, I would be open to adding to this PR once decided on the ordering method.
> 

As a user, what ordering would make more sense to you? Or just alphabetically would be fine, as at least it's easier to know what should come next?

**Comment by jmtaylor90** [comment]
> > > This was massive, thanks for this undertaking!
> > > Left mostly nit comments for typos and the likes., not very confident to review HTTP workings or rule keyword details...
> > > I wonder if something can be done about the sections that have several notes, as I worry that the sheer amount may render some less visible.
> > 
> > 
> > Indeed, some of the keywords just have a number of details that are important. Without making the document a wall of text, I figured using the note:: option at least broke up the content a little but am open to suggestions on that
> 
> Unfortunately, no suggestions, for now, might be the easiest solution, indeed.
> 
> > > Once wondering if we could reorganize the keyword order alphabetically, or semantically, somehow (might be a separate task, if agreed upon).
> > 
> > 
> > I was wondering about ordering as well, I would be open to adding to this PR once decided on the ordering method.
> 
> As a user, what ordering would make more sense to you? Or just alphabetically would be fine, as at least it's easier to know what should come next?

I will have a new PR next week with the alphabetical ordering and the changes suggested so far. Thanks!

File: doc/userguide/rules/http-keywords.rst [hunk 13]
```diff
@@ -1,813 +1,1200 @@
...
-Notes
-~~~~~
+.. note:: ``http.user_agent`` does not include the leading space or trailing
+   \\r\\n
 
--  The ``http.user_agent`` buffer will NOT include the header name,
-   colon, or leading whitespace.  i.e. it will not include
-   "User-Agent: ".
+.. note:: Using the ``http.user_agent`` buffer is more efficient when it comes
+  to performance than using the :ref:`http.header` buffer (~10% better).
```
[the remark this answers is not in the document]
Comment (jmtaylor90) [reply doc/userguide/rules/http-keywords.rst:426]: I am not sure but if we don't it's probably worth removing that performance note.

[the remark this answers is not in the document]
Comment (jufajardini) [reply doc/userguide/rules/http-keywords.rst:426]: Aha, it comes from this: https://blog.inliniac.net/2012/07/09/suricata-http_user_agent-vs-http_header/

[the remark this answers is not in the document]
Comment (jmtaylor90) [reply doc/userguide/rules/http-keywords.rst:426]: Do we think that is still accurate enough to mention the performance difference?

[the remark this answers is not in the document]
Comment (jufajardini) [reply doc/userguide/rules/http-keywords.rst:426]: Checked with the rest of the team

[the remark this answers is not in the document]
Comment (jmtaylor90) [reply doc/userguide/rules/http-keywords.rst:426]: what was the verdict? Keep it in or ?

[the remark this answers is not in the document]
Comment (jufajardini) [reply doc/userguide/rules/http-keywords.rst:426]: So, according to Victor: It's inspecting less data, so it should be faster. But it's a bit less clear by how much.
Maybe we could add that as it inspects data, it is faster? Do you think that'd work?

[the remark this answers is not in the document]
Comment (jmtaylor90) [reply doc/userguide/rules/http-keywords.rst:426]: Sounds good, will rework it. Thanks!

**Comment by jmtaylor90** [comment]
continued in https://github.com/OISF/suricata/pull/10599

**Pull request closed without merging**